### PR TITLE
[ENH] Add nan policy handler for FallbackForecaster #5897

### PR DESCRIPTION
**Pull request by ninedigits**
#### Reference Issues/PRs
-Issue: https://github.com/sktime/sktime/issues/5897
-Additional discussion: https://github.com/sktime/sktime/pull/5847

#### What does this implement/fix? Explain your changes.

We've introduced a `nan_predict_policy` within the `FallbackForecaster` class to handle NaN values in forecaster predictions. This policy defaults to "ignore" and can be set to "raise", which will actively check for NaN values in the prediction output and raise an exception if any are detected, or "warn", which will warn the user if any values are detected. This enhancement is aimed at improving the robustness and reliability of the forecasting process, especially when integrating transformations like BoxCox with ARIMA forecasters.

#### Does your contribution introduce a new dependency? If yes, which one?

No new dependencies.

#### What should a reviewer concentrate their feedback on?

- Correctness of the NaN detection and exception raising mechanism.
- Adhering to existing `sktime` design standards.
- Adequacy of tests covering the new functionality.

#### Did you add any tests for the change?

Yes, tests have been added to ensure the `nan_predict_policy` behaves as expected in scenarios where NaN values are predicted. These tests cover both the "raise" and "warn" policies to verify their correct implementation.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

File: sktime/forecasting/compose/_fallback.py [hunk 1]
```diff
@@ -36,6 +47,13 @@ class is useful in scenarios where the reliability of individual forecasting mod
     verbose : bool, default=False
         If True, raises warnings when a forecaster fails to fit or predict.
 
+    nan_predict_policy: str, default='ignore'
+        Determines the action to take if NaN values are found in the predictions. If set
+         to 'raise', the method raises a ValueError when NaN values are detected in the
```
Comment (fkiraly) [inline sktime/forecasting/compose/_fallback.py:52]: is this indentation change intented? Might render strangely.

Comment (fkiraly) [inline sktime/forecasting/compose/_fallback.py:52]: I would explain `raise` better - the exception is in fact not raised to hte user, but handled by this class, i.e., occurrence of nans is considered a failure for the fallback

File: sktime/forecasting/compose/_fallback.py [hunk 2]
```diff
@@ -96,14 +114,17 @@ class is useful in scenarios where the reliability of individual forecasting mod
...
+    def __init__(self, forecasters, verbose=False, nan_predict_policy="ignore"):
         super().__init__()
 
         self.forecasters = forecasters
         self.current_forecaster_ = None
         self.current_name_ = None
         self.verbose = verbose
 
+        _check_nan_policy_option(nan_predict_policy)
+        self.nan_predict_policy = nan_predict_policy
```
Comment (fkiraly) [inline sktime/forecasting/compose/_fallback.py:126]: slightly more idiomatic (similar ot sklearn): have the checker return the input, then do `self.palim = check_palim(palim)`

File: sktime/forecasting/compose/_fallback.py [hunk 3]
```diff
@@ -112,6 +133,20 @@ def __init__(self, forecasters, verbose=False):
         self._anytagis_then_set("requires-fh-in-fit", True, False, self._forecasters)
         self._anytagis_then_set("capability:pred_int", False, True, self._forecasters)
 
+    def _validate_y_pred(self, y_pred):
+        if self.nan_predict_policy in ("warn", "raise"):
+            num_nans = y_pred.isna().sum()
```
Comment (fkiraly) [inline sktime/forecasting/compose/_fallback.py:138]: * should this use `isnull`? Or should it be an option to use `isna` vs `isnull`?
* `any` is faster than `sum` in this case

**Review by fkiraly — CHANGES_REQUESTED** [review]
Seems to work, and it is tested.

I've added some comments, but would consider none of these blocking.

File: sktime/forecasting/compose/_fallback.py [hunk 4]
```diff
@@ -112,6 +133,20 @@ def __init__(self, forecasters, verbose=False):
...
+            if num_nans > 0:
+                msg = f"Null value presents in predict: {y_pred}"
+                if self.nan_predict_policy == "raise":
```
[the remark this answers is not in the document]
Comment (ninedigits) [reply sktime/forecasting/compose/_fallback.py:141]: @fkiraly Isn't this handled in line 137?
```python3
    def _validate_y_pred(self, y_pred):
        if self.nan_predict_policy in ("warn", "raise"):
            num_nans = y_pred.isna().sum() ### todo, change this to isnull().any()
```

[the remark this answers is not in the document]
Comment (fkiraly) [reply sktime/forecasting/compose/_fallback.py:141]: oh, yes, it is.

File: sktime/forecasting/compose/_fallback.py (hunk 1 above)
Comment (ninedigits) [reply sktime/forecasting/compose/_fallback.py:52]: Good catch

**Comment by ninedigits** [comment]
> Seems to work, and it is tested.
> 
> I've added some comments, but would consider none of these blocking.

Good revisions suggestions, I added them in.

File: sktime/forecasting/compose/_fallback.py (hunk 1 above)
Comment (fkiraly) [reply sktime/forecasting/compose/_fallback.py]: I mean, is the intentation indented, no, the intendation intented - well, you know what I mean

Comment (ninedigits) [reply sktime/forecasting/compose/_fallback.py]: I understood what you meant, the indentation was wrong. Was a good catch!

**Comment by ninedigits** [comment]
In the future, we can add to `_validate_y_pred()` to handle other non-exception forecasting errors. E.g. exploding forecasts, although that one could be tricky to implement.

File: sktime/forecasting/compose/_fallback.py (hunk 1 above)
Comment (fkiraly) [reply sktime/forecasting/compose/_fallback.py]: I was just joking about the typo I made

File: sktime/forecasting/compose/_fallback.py [hunk 5]
```diff
@@ -36,6 +48,22 @@ class is useful in scenarios where the reliability of individual forecasting mod
     verbose : bool, default=False
         If True, raises warnings when a forecaster fails to fit or predict.
 
+    nan_predict_policy: str, default='ignore'
+        Determines the action to take if NaN values are found in the predictions.
+        Available options:
+        * "ignore"
```
Comment (fkiraly) [inline sktime/forecasting/compose/_fallback.py:54]: I think this does not render right if there is no space before and after the bullet point list

**Review by fkiraly — APPROVED** [review]
Thanks - one non-blocking comment above about a formatting problem in the docstring.

**Comment by fkiraly** [comment]
> In the future, we can add to `_validate_y_pred()` to handle other non-exception forecasting errors. E.g. exploding forecasts, although that one could be tricky to implement.

Very interesting idea. Perhaps it is worthwhile to think about how to make this abstract? E.g., allow any parameter estimator to define the condition? Another place where sth like this exists is the `TransformIf` transformer, so I wonder whether we can split off a "condition" object or parameter estimator?

**Comment by ninedigits** [comment]
> Thanks - one non-blocking comment above about a formatting problem in the docstring.

Committed those changes

**Comment by ninedigits** [comment]
> > In the future, we can add to `_validate_y_pred()` to handle other non-exception forecasting errors. E.g. exploding forecasts, although that one could be tricky to implement.
> 
> Very interesting idea. Perhaps it is worthwhile to think about how to make this abstract? E.g., allow any parameter estimator to define the condition? Another place where sth like this exists is the `TransformIf` transformer, so I wonder whether we can split off a "condition" object or parameter estimator?

Like a parameter where a function or estimator is passed in capable of assessing the inputs and outputs, and allowing the user to specify their own validation params?

**Comment by fkiraly** [comment]
> Like a parameter where a function or estimator is passed in capable of assessing the inputs and outputs, and allowing the user to specify their own validation params?

Yes! The closest type of such estimator/object is the "parameter estimator", currently used mostly for stationarity and seasonality detection, the scitype code is `param_est` (for use in `all_estimators`). New issue to discuss?

**Review by fkiraly — APPROVED** [review]

**Comment by ninedigits** [comment]
> Yes! The closest type of such estimator/object is the "parameter estimator", currently used mostly for stationarity and seasonality detection, the scitype code is `param_est` (for use in `all_estimators`). New issue to discuss?

It's a good idea, I can take a look at it. I'm wondering if it would be it's own estimator that would wrap around an input forecaster within FallbackForecaster?

**Comment by fkiraly** [comment]
good question - I think there's a deeper API design discussion her.e

Opened an issue to move it out of this PR which will get merged soon: https://github.com/sktime/sktime/issues/5936

Simple reason, merged PR are very low visibility, whereas open issues signify ongoing work.

**Pull request merged**